In [74]:
import pandas
import plotly.express as px

<h1 style="text-align:center">⛏ Data explorations</h1> 

# 1. Prices data from the [Binance API](https://binance-docs.github.io/apidocs/spot/en/#kline-candlestick-data) 💰

In [75]:
prices = pandas.read_csv('../../datastore/prices.csv')
prices

Symbol  base quote  Kline open time  Open price  High price  \
0         ETHBTC   ETH   BTC    1669395600000    0.072026    0.072222   
1         ETHBTC   ETH   BTC    1669397400000    0.072222    0.072335   
2         ETHBTC   ETH   BTC    1669399200000    0.072163    0.072263   
3         ETHBTC   ETH   BTC    1669401000000    0.072188    0.072261   
4         ETHBTC   ETH   BTC    1669402800000    0.072223    0.072479   
...          ...   ...   ...              ...         ...         ...   
1222237  BUSDRON  BUSD   RON    1670943600000    4.629000    4.630000   
1222238  BUSDRON  BUSD   RON    1670945400000    4.620000    4.631000   
1222239  BUSDRON  BUSD   RON    1670947200000    4.631000    4.631000   
1222240  BUSDRON  BUSD   RON    1670949000000    4.630000    4.631000   
1222241  BUSDRON  BUSD   RON    1670950800000    4.631000    4.631000   

         Low price  Close price      Volume  Kline Close time  \
0         0.072026     0.072222   1342.1624     1669397399999   
1         0.072117     0.072162   1185.9640     1669399199999   
2         0.072101     0.072189    813.4793     1669400999999   
3         0.072161     0.072212    936.2168     1669402799999   
4         0.072197     0.072375   1115.5725     1669404599999   
...            ...          ...         ...               ...   
1222237   4.620000     4.621000   5984.0000     1670945399999   
1222238   4.620000     4.630000   3618.0000     1670947199999   
1222239   4.630000     4.631000   3254.0000     1670948999999   
1222240   4.630000     4.631000  14415.0000     1670950799999   
1222241   4.630000     4.631000   3875.0000     1670952599999   

         Quote asset volume  Number of trades  Taker buy base asset volume  \
0                 96.776126              2657                     794.6663   
1                 85.640393              2727                     560.9044   
2                 58.717779              1862                     359.8931   
3                 67.599402              1816                     470.6928   
4                 80.727099              3088                     593.4310   
...                     ...               ...                          ...   
1222237        27672.337000               108                    2086.0000   
1222238        16734.982000                42                    1642.0000   
1222239        15069.018000                33                    2998.0000   
1222240        66749.027000                92                    7577.0000   
1222241        17943.045000                52                    1795.0000   

         Taker buy quote asset volume  
0                           57.298344  
1                           40.508977  
2                           25.975732  
3                           33.988586  
4                           42.945939  
...                               ...  
1222237                   9651.160000  
1222238                   7593.482000  
1222239                  13883.738000  
1222240                  35089.087000  
1222241                   8312.645000  

[1222242 rows x 14 columns]

## 💎 What is the meaning of each column?
* symbol : mã kí hiệu loại giao dịch
* base : mã đồng cơ sở
* quote : mã đồng báo giá
* Kline open/close time : thời gian mở/đóng của việc thu dữ liệu
* open price : Giá tại thời điểm đầu khi thu dữ liệu
* close price : giá tại thời điểm cuối cùng khi thu dữ liệu
* high price : giá cao nhất trong thời gian thu dữ liệu
* low price : giá thấp nhất trong thời gian thu dữ liệu
* volume : số lượng base đã được trao đổi
* quote asset volume : số lượng quote đã được trao đổi
* taker buy base volume : số lượng mua base (lệnh) (?)
* taker buy quote volume : số lượng mua quote (lệnh) (?)

## 💎 What is the current data type of each column?

In [76]:
prices.dtypes

Symbol                           object
base                             object
quote                            object
Kline open time                   int64
Open price                      float64
High price                      float64
Low price                       float64
Close price                     float64
Volume                          float64
Kline Close time                  int64
Quote asset volume              float64
Number of trades                  int64
Taker buy base asset volume     float64
Taker buy quote asset volume    float64
dtype: object

<p style="margin-left:1rem">❓ Which columns have missing data</p> 

In [77]:
for column in prices.columns:
    print("Missing values in column {}: {} ({}%)".format(column, prices[column].isnull().sum(), prices[column].isnull().sum() / len(prices) * 100))

Missing values in column Symbol: 0 (0.0%)
Missing values in column base: 0 (0.0%)
Missing values in column quote: 0 (0.0%)
Missing values in column Kline open time: 0 (0.0%)
Missing values in column Open price: 0 (0.0%)
Missing values in column High price: 0 (0.0%)
Missing values in column Low price: 0 (0.0%)
Missing values in column Close price: 0 (0.0%)
Missing values in column Volume: 0 (0.0%)
Missing values in column Kline Close time: 0 (0.0%)
Missing values in column Quote asset volume: 0 (0.0%)
Missing values in column Number of trades: 0 (0.0%)
Missing values in column Taker buy base asset volume: 0 (0.0%)
Missing values in column Taker buy quote asset volume: 0 (0.0%)


<p style="margin-left:2rem">➡️ Formatting open and close time</p> 

In [78]:
prices["Kline open time"] = pandas.to_datetime(prices["Kline open time"], unit='ms')
prices["Kline Close time"] = pandas.to_datetime(prices["Kline Close time"], unit='ms')

<p style="margin-left:1rem">❓ Are there any 2 symbols duplicated (USDTBTC is also a duplicated symbol of BTCUSDT)</p> 

<p style="margin-left:2rem">➡️ Drop symbols have the same name</p> 

In [79]:
drop_dupl_symbol_df = prices.drop_duplicates(subset='Symbol')
drop_dupl_symbol_df

Symbol   base quote     Kline open time  Open price  High price  \
0           ETHBTC    ETH   BTC 2022-11-25 17:00:00    0.072026    0.072222   
865         LTCBTC    LTC   BTC 2022-11-25 17:00:00    0.004513    0.004537   
1730        BNBBTC    BNB   BTC 2022-11-25 17:00:00    0.018190    0.018290   
2595        NEOBTC    NEO   BTC 2022-11-25 17:00:00    0.000415    0.000416   
3460       QTUMETH   QTUM   ETH 2022-11-25 17:00:00    0.001864    0.001864   
...            ...    ...   ...                 ...         ...         ...   
1221396    HOOKBNB   HOOK   BNB 2022-12-01 14:30:00    0.000338    0.020000   
1221978   MAGICBTC  MAGIC   BTC 2022-12-12 08:00:00    0.000022    0.000098   
1222045  MAGICBUSD  MAGIC  BUSD 2022-12-12 08:00:00    0.382200    0.924000   
1222112  MAGICUSDT  MAGIC  USDT 2022-12-12 08:00:00    0.382100    0.902900   
1222179    BUSDRON   BUSD   RON 2022-12-12 10:00:00    4.680000    4.680000   

         Low price  Close price        Volume        Kline Close time  \
0         0.072026     0.072222  1.342162e+03 2022-11-25 17:29:59.999   
865       0.004513     0.004523  1.973337e+03 2022-11-25 17:29:59.999   
1730      0.018189     0.018223  1.035607e+03 2022-11-25 17:29:59.999   
2595      0.000415     0.000416  2.472300e+02 2022-11-25 17:29:59.999   
3460      0.001864     0.001864  2.600000e+00 2022-11-25 17:29:59.999   
...            ...          ...           ...                     ...   
1221396   0.000338     0.006934  3.782472e+06 2022-12-01 14:59:59.999   
1221978   0.000022     0.000053  1.154971e+06 2022-12-12 08:29:59.999   
1222045   0.382200     0.894900  1.520490e+07 2022-12-12 08:29:59.999   
1222112   0.382100     0.890000  1.494773e+07 2022-12-12 08:29:59.999   
1222179   4.640000     4.667000  2.133500e+04 2022-12-12 10:29:59.999   

         Quote asset volume  Number of trades  Taker buy base asset volume  \
0              9.677613e+01              2657                 7.946663e+02   
865            8.928513e+00               718                 9.773160e+02   
1730           1.888992e+01              1762                 4.612330e+02   
2595           1.027427e-01                33                 1.480600e+02   
3460           4.846400e-03                 1                 2.600000e+00   
...                     ...               ...                          ...   
1221396        2.634135e+04             32175                 1.738188e+06   
1221978        5.548904e+01              5915                 4.258004e+05   
1222045        1.248077e+07             58295                 6.943390e+06   
1222112        1.218384e+07             38714                 8.425779e+06   
1222179        9.924754e+04               122                 6.713000e+03   

         Taker buy quote asset volume  
0                        5.729834e+01  
865                      4.422118e+00  
1730                     8.413329e+00  
2595                     6.153731e-02  
3460                     4.846400e-03  
...                               ...  
1221396                  1.234107e+04  
1221978                  2.052682e+01  
1222045                  5.707546e+06  
1222112                  6.871771e+06  
1222179                  3.132106e+04  

[1418 rows x 14 columns]

<p style="margin-left:2rem">➡️ Finding symbols have the inverse name (BTCUSDT and USDTBTC) => There are no symbols </p> 

In [80]:
duplicate_symbol = []

def find_dupl(row):
    matches = drop_dupl_symbol_df.apply(lambda x: x['base'] == row['quote'] and x['quote'] == row['base'], axis=1)
    if matches.any():
        duplicate_symbol.append(row['Symbol'])
    return row
    

drop_dupl_symbol_df.apply(find_dupl, axis=1)

print("Duplicate symbols: ", duplicate_symbol)

Duplicate symbols:  []


## 💎 With each column, how are values distributed?

In [81]:
prices.describe()

Open price    High price     Low price   Close price        Volume  \
count  1.222242e+06  1.222242e+06  1.222242e+06  1.222242e+06  1.222242e+06   
mean   2.071786e+05  2.075591e+05  2.067964e+05  2.071887e+05  8.827653e+07   
std    7.071014e+06  7.083959e+06  7.058199e+06  7.071409e+06  3.045814e+09   
min    1.900000e-07  1.900000e-07  1.900000e-07  1.900000e-07  0.000000e+00   
25%    4.353225e-04  4.372150e-04  4.344000e-04  4.352050e-04  3.726848e+02   
50%    9.240000e-02  9.280000e-02  9.200000e-02  9.240000e-02  6.638850e+03   
75%    1.772000e+00  1.781000e+00  1.764000e+00  1.772000e+00  7.157023e+04   
max    2.800114e+08  2.952054e+08  2.795345e+08  2.802426e+08  1.087945e+12   

       Quote asset volume  Number of trades  Taker buy base asset volume  \
count        1.222242e+06      1.222242e+06                 1.222242e+06   
mean         1.982917e+07      3.202449e+02                 4.355778e+07   
std          7.413260e+09      3.337801e+03                 1.606299e+09   
min          0.000000e+00      0.000000e+00                 0.000000e+00   
25%          4.566503e-01      8.000000e+00                 7.683125e+01   
50%          1.922798e+03      3.500000e+01                 2.428761e+03   
75%          2.118361e+04      1.260000e+02                 3.014231e+04   
max          4.766742e+12      4.657260e+05                 6.547297e+11   

       Taker buy quote asset volume  
count                  1.222242e+06  
mean                   1.005756e+07  
std                    3.704048e+09  
min                    0.000000e+00  
25%                    9.237323e-02  
50%                    6.406491e+02  
75%                    9.344149e+03  
max                    2.373518e+12

<p style="margin-left:2rem">➡️ Each symbol with have 865 rows of data, each row represents a 1 minute interval and there are totally 1418 symbols</p> 

In [82]:
prices.groupby("Symbol").count()[["Kline open time"]]

Kline open time
Symbol                    
1INCHBTC               865
1INCHBUSD              865
1INCHUSDT              865
AAVEBNB                865
AAVEBTC                865
...                    ...
ZILUSDT                865
ZRXBTC                 865
ZRXBUSD                865
ZRXETH                 865
ZRXUSDT                865

[1418 rows x 1 columns]

<p style="margin-left:2rem">➡️ Open and close time columns</p> 


In [83]:
print("Kline open time: ")
print(prices["Kline open time"].min())
print(prices["Kline open time"].max())
print("Kline close time: ")
print(prices["Kline Close time"].min())
print(prices["Kline Close time"].max())

Kline open time: 
2022-11-25 17:00:00
2022-12-13 17:00:00
Kline close time: 
2022-11-25 17:29:59.999000
2022-12-13 17:29:59.999000


# 2. Symbol data from the [Coin Market Cap API](https://coinmarketcap.com/api/documentation/v1/#operation/getV1CryptocurrencyListingsLatest) 📉

In [100]:
symbols = pandas.read_csv('../../datastore/coin_market_cap_symbols.csv')
symbols

symbol         price  percent_change_7d    market_cap  \
0       BTC  16651.011406          -1.327824  3.204503e+11   
1       ETH   1195.844971          -1.641379  1.463402e+11   
2      USDT      0.999861          -0.026345  6.623844e+10   
3      USDC      1.000009          -0.008786  4.422006e+10   
4       BNB    243.613912          -2.177910  3.896979e+10   
...     ...           ...                ...           ...   
4995   ZEUS      0.078431         -16.032448  0.000000e+00   
4996    JDC      0.037161           6.302088  0.000000e+00   
4997   LIQD   1067.949442          -0.318138  0.000000e+00   
4998    PEX      0.000363          -0.590274  0.000000e+00   
4999  BHIVE      0.272204          -3.281791  0.000000e+00   

                                                   tags  \
0     mineable,pow,sha-256,store-of-value,state-chan...   
1     pos,smart-contracts,ethereum-ecosystem,coinbas...   
2     payments,stablecoin,asset-backed-stablecoin,av...   
3     medium-of-exchange,stablecoin,asset-backed-sta...   
4     marketplace,centralized-exchange,payments,smar...   
...                                                 ...   
4995                                                NaN   
4996                                                NaN   
4997                                  arbitrum-ecosytem   
4998                                                NaN   
4999                                                NaN   

                    date_added  
0     2013-04-28T00:00:00.000Z  
1     2015-08-07T00:00:00.000Z  
2     2015-02-25T00:00:00.000Z  
3     2018-10-08T00:00:00.000Z  
4     2017-07-25T00:00:00.000Z  
...                        ...  
4995  2022-10-07T08:12:00.000Z  
4996  2019-12-06T00:00:00.000Z  
4997  2022-08-29T01:16:10.000Z  
4998  2021-11-26T00:31:12.000Z  
4999  2022-10-03T04:29:14.000Z  

[5000 rows x 6 columns]

## 💎 What is the meaning of each column?
* symbol: symbol of the coin
* price: price of the symbol by USD
* percent_change_7d: percent change of the price in 7 days
* market_cap: market capitalization of the symbol by USD
* tags: tags of the symbol
* date_added: date added to the coinmarketcap

## 💎 What is the current data type of each column?

In [85]:
symbols.dtypes

symbol                object
price                float64
percent_change_7d    float64
market_cap           float64
tags                  object
date_added            object
dtype: object

<p style="margin-left:1rem">❓ Which columns have missing data</p> 

In [86]:
for column in symbols.columns:
    print("Missing values in column {}: {} ({}%)".format(column, symbols[column].isnull().sum(), symbols[column].isnull().sum() / len(symbols) * 100))

Missing values in column symbol: 0 (0.0%)
Missing values in column price: 0 (0.0%)
Missing values in column percent_change_7d: 0 (0.0%)
Missing values in column market_cap: 0 (0.0%)
Missing values in column tags: 2307 (46.14%)
Missing values in column date_added: 0 (0.0%)


<p style="margin-left:2rem">➡️ Filling missing data</p> 

In [87]:
symbols["tags"] = symbols["tags"].fillna("")

<p style="margin-left:2rem">➡️ Format date column</p> 

In [88]:
symbols["date_added"] = pandas.to_datetime(symbols["date_added"])

<p style="margin-left:2rem">➡️ Data after formatted</p> 

In [89]:
symbols

symbol         price  percent_change_7d    market_cap  \
0       BTC  16651.011406          -1.327824  3.204503e+11   
1       ETH   1195.844971          -1.641379  1.463402e+11   
2      USDT      0.999861          -0.026345  6.623844e+10   
3      USDC      1.000009          -0.008786  4.422006e+10   
4       BNB    243.613912          -2.177910  3.896979e+10   
...     ...           ...                ...           ...   
4995   ZEUS      0.078431         -16.032448  0.000000e+00   
4996    JDC      0.037161           6.302088  0.000000e+00   
4997   LIQD   1067.949442          -0.318138  0.000000e+00   
4998    PEX      0.000363          -0.590274  0.000000e+00   
4999  BHIVE      0.272204          -3.281791  0.000000e+00   

                                                   tags  \
0     mineable,pow,sha-256,store-of-value,state-chan...   
1     pos,smart-contracts,ethereum-ecosystem,coinbas...   
2     payments,stablecoin,asset-backed-stablecoin,av...   
3     medium-of-exchange,stablecoin,asset-backed-sta...   
4     marketplace,centralized-exchange,payments,smar...   
...                                                 ...   
4995                                                      
4996                                                      
4997                                  arbitrum-ecosytem   
4998                                                      
4999                                                      

                    date_added  
0    2013-04-28 00:00:00+00:00  
1    2015-08-07 00:00:00+00:00  
2    2015-02-25 00:00:00+00:00  
3    2018-10-08 00:00:00+00:00  
4    2017-07-25 00:00:00+00:00  
...                        ...  
4995 2022-10-07 08:12:00+00:00  
4996 2019-12-06 00:00:00+00:00  
4997 2022-08-29 01:16:10+00:00  
4998 2021-11-26 00:31:12+00:00  
4999 2022-10-03 04:29:14+00:00  

[5000 rows x 6 columns]

In [90]:
symbols.dtypes

symbol                            object
price                            float64
percent_change_7d                float64
market_cap                       float64
tags                              object
date_added           datetime64[ns, UTC]
dtype: object

<p style="margin-left:1rem">❓ Are there any 2 symbols duplicated (USDTBTC is also a duplicated symbol of BTCUSDT)</p> 

In [98]:
symbols = symbols.drop_duplicates("symbol")
symbols

symbol         price  percent_change_7d    market_cap  \
0       BTC  16651.011406          -1.327824  3.204503e+11   
1       ETH   1195.844971          -1.641379  1.463402e+11   
2      USDT      0.999861          -0.026345  6.623844e+10   
3      USDC      1.000009          -0.008786  4.422006e+10   
4       BNB    243.613912          -2.177910  3.896979e+10   
...     ...           ...                ...           ...   
4992    FEY      0.005847          10.285887  0.000000e+00   
4995   ZEUS      0.078431         -16.032448  0.000000e+00   
4996    JDC      0.037161           6.302088  0.000000e+00   
4997   LIQD   1067.949442          -0.318138  0.000000e+00   
4999  BHIVE      0.272204          -3.281791  0.000000e+00   

                                                   tags  \
0     [mineable, pow, sha-256, store-of-value, state...   
1     [pos, smart-contracts, ethereum-ecosystem, coi...   
2     [payments, stablecoin, asset-backed-stablecoin...   
3     [medium-of-exchange, stablecoin, asset-backed-...   
4     [marketplace, centralized-exchange, payments, ...   
...                                                 ...   
4992                                                 []   
4995                                                 []   
4996                                                 []   
4997                                [arbitrum-ecosytem]   
4999                                                 []   

                    date_added  
0    2013-04-28 00:00:00+00:00  
1    2015-08-07 00:00:00+00:00  
2    2015-02-25 00:00:00+00:00  
3    2018-10-08 00:00:00+00:00  
4    2017-07-25 00:00:00+00:00  
...                        ...  
4992 2021-06-09 00:00:00+00:00  
4995 2022-10-07 08:12:00+00:00  
4996 2019-12-06 00:00:00+00:00  
4997 2022-08-29 01:16:10+00:00  
4999 2022-10-03 04:29:14+00:00  

[4630 rows x 6 columns]

## 💎 With each column, how are values distributed?

In [99]:
symbols.describe()

price  percent_change_7d    market_cap
count  4.630000e+03        4630.000000  4.630000e+03
mean   1.723140e+02           0.649045  1.791379e+08
std    5.490075e+03          59.499285  5.359677e+09
min    3.069622e-18         -99.995832  0.000000e+00
25%    1.140524e-03          -7.220396  0.000000e+00
50%    1.122148e-02          -1.891207  1.625297e+04
75%    1.413062e-01           1.255025  1.676742e+06
max    3.584306e+05        3132.463931  3.204503e+11

<p style="margin-left:2rem">➡️ The date that symbols are added</p> 

In [101]:
print("Date added: ")
print(symbols["date_added"].min())
print(symbols["date_added"].max())

Date added: 
2013-04-28T00:00:00.000Z
2022-12-28T05:27:34.000Z


<p style="margin-left:2rem">➡️ The most popular tag has attach to 2307 symbols</p> 

In [102]:
symbols["tags"] = symbols["tags"].apply(lambda x: str(x).split(","))
symbols_with_tags_df = symbols.explode("tags")
symbols_with_tags_df

symbol         price  percent_change_7d    market_cap               tags  \
0       BTC  16651.011406          -1.327824  3.204503e+11           mineable   
0       BTC  16651.011406          -1.327824  3.204503e+11                pow   
0       BTC  16651.011406          -1.327824  3.204503e+11            sha-256   
0       BTC  16651.011406          -1.327824  3.204503e+11     store-of-value   
0       BTC  16651.011406          -1.327824  3.204503e+11      state-channel   
...     ...           ...                ...           ...                ...   
4995   ZEUS      0.078431         -16.032448  0.000000e+00                nan   
4996    JDC      0.037161           6.302088  0.000000e+00                nan   
4997   LIQD   1067.949442          -0.318138  0.000000e+00  arbitrum-ecosytem   
4998    PEX      0.000363          -0.590274  0.000000e+00                nan   
4999  BHIVE      0.272204          -3.281791  0.000000e+00                nan   

                    date_added  
0     2013-04-28T00:00:00.000Z  
0     2013-04-28T00:00:00.000Z  
0     2013-04-28T00:00:00.000Z  
0     2013-04-28T00:00:00.000Z  
0     2013-04-28T00:00:00.000Z  
...                        ...  
4995  2022-10-07T08:12:00.000Z  
4996  2019-12-06T00:00:00.000Z  
4997  2022-08-29T01:16:10.000Z  
4998  2021-11-26T00:31:12.000Z  
4999  2022-10-03T04:29:14.000Z  

[11127 rows x 6 columns]

In [103]:
grouped_tags = symbols_with_tags_df.groupby("tags").count()[["symbol"]]
grouped_tags.describe()

symbol
count   289.000000
mean     38.501730
std     146.689916
min       1.000000
25%       3.000000
50%      11.000000
75%      34.000000
max    2307.000000

<p style="margin-left:1rem">❓ Which symbols are priced close to USD and are reliable to store coins on the wallet</p> 

In [104]:
stable_coin_df = symbols.loc[(symbols['percent_change_7d'].abs() < 0.1) & ((symbols['price'].astype(float) - 1).abs() < 0.01)]
stable_coin_df

symbol     price  percent_change_7d    market_cap  \
2          USDT  0.999861          -0.026345  6.623844e+10   
3          USDC  1.000009          -0.008786  4.422006e+10   
6          BUSD  0.999857          -0.035373  1.717318e+10   
10          DAI  0.999559          -0.037205  5.814973e+09   
44         TUSD  0.999991           0.072637  7.542295e+08   
331        OUSD  0.997073          -0.072119  4.791342e+07   
362        MUSD  0.994568           0.000000  4.086279e+07   
2593  BB-A-USDT  1.004441          -0.012645  0.000000e+00   
2594       EDGT  1.000353           0.040720  0.000000e+00   
2596  BB-A-USDC  1.001356           0.015670  0.000000e+00   
2612      XTUSD  1.000790           0.069436  0.000000e+00   
2664        BOB  0.999858           0.005153  0.000000e+00   
2667       USDZ  0.996965          -0.054847  0.000000e+00   
2669    MIMATIC  0.992428          -0.030816  0.000000e+00   
2707       xDAI  1.000215          -0.010686  0.000000e+00   
2724      wxDai  1.000174          -0.012186  0.000000e+00   
3121        MIM  0.998102           0.026359  0.000000e+00   
3172       UUSD  0.994611          -0.086657  0.000000e+00   
3433      VEUSD  1.000118           0.037879  0.000000e+00   
3839    axlUSDC  0.999960          -0.002185  0.000000e+00   
4137    WANUSDT  1.000012           0.011087  0.000000e+00   

                                                   tags  \
2     [payments, stablecoin, asset-backed-stablecoin...   
3     [medium-of-exchange, stablecoin, asset-backed-...   
6     [stablecoin, asset-backed-stablecoin, binance-...   
10    [defi, stablecoin, asset-backed-stablecoin, et...   
44    [store-of-value, stablecoin, asset-backed-stab...   
331                  [defi, stablecoin, usd-stablecoin]   
362   [defi, stablecoin, asset-backed-stablecoin, po...   
2593                                              [nan]   
2594                                              [nan]   
2596                                              [nan]   
2612                                              [nan]   
2664                                              [nan]   
2667                                       [stablecoin]   
2669  [stablecoin, algorithmic-stablecoin, arbitrum-...   
2707                              [scaling, stablecoin]   
2724                                              [nan]   
3121  [mineable, stablecoin, avalanche-ecosystem, ar...   
3172                                              [nan]   
3433                                              [nan]   
3839                                              [nan]   
4137                                       [stablecoin]   

                    date_added  
2     2015-02-25T00:00:00.000Z  
3     2018-10-08T00:00:00.000Z  
6     2019-09-20T00:00:00.000Z  
10    2019-11-22T00:00:00.000Z  
44    2018-03-06T00:00:00.000Z  
331   2020-09-27T00:00:00.000Z  
362   2020-06-28T00:00:00.000Z  
2593  2022-05-30T12:50:12.000Z  
2594  2021-04-27T00:00:00.000Z  
2596  2022-05-30T12:50:12.000Z  
2612  2022-07-18T09:47:29.000Z  
2664  2022-09-21T02:15:19.000Z  
2667  2022-06-06T15:25:17.000Z  
2669  2021-06-02T00:00:00.000Z  
2707  2021-03-01T00:00:00.000Z  
2724  2021-03-30T00:00:00.000Z  
3121  2014-03-12T00:00:00.000Z  
3172  2022-11-17T10:36:36.000Z  
3433  2022-01-04T19:52:04.000Z  
3839  2022-08-17T11:09:01.000Z  
4137  2021-03-02T00:00:00.000Z

<p style="margin-left:1rem">❓ Which coin is available for the most trading</p> 

In [105]:
f_data = prices.loc[prices['quote'].isin(stable_coin_df.symbol) & (~prices['base'].isin(stable_coin_df.symbol))] # symbols which has stable-coin quote
f_data

Symbol   base quote     Kline open time  Open price  High price  \
8650       BTCUSDT    BTC  USDT 2022-11-25 17:00:00  16485.1400  16512.4300   
8651       BTCUSDT    BTC  USDT 2022-11-25 17:30:00  16483.2700  16536.3500   
8652       BTCUSDT    BTC  USDT 2022-11-25 18:00:00  16509.5200  16528.6000   
8653       BTCUSDT    BTC  USDT 2022-11-25 18:30:00  16508.9500  16518.5900   
8654       BTCUSDT    BTC  USDT 2022-11-25 19:00:00  16495.4300  16553.8800   
...            ...    ...   ...                 ...         ...         ...   
1222174  MAGICUSDT  MAGIC  USDT 2022-12-13 15:00:00      0.7394      0.7467   
1222175  MAGICUSDT  MAGIC  USDT 2022-12-13 15:30:00      0.7156      0.7166   
1222176  MAGICUSDT  MAGIC  USDT 2022-12-13 16:00:00      0.7090      0.7216   
1222177  MAGICUSDT  MAGIC  USDT 2022-12-13 16:30:00      0.7176      0.7180   
1222178  MAGICUSDT  MAGIC  USDT 2022-12-13 17:00:00      0.7149      0.7171   

          Low price  Close price        Volume        Kline Close time  \
8650     16474.2900   16483.2700    3369.76766 2022-11-25 17:29:59.999   
8651     16482.9800   16510.3800    3892.49531 2022-11-25 17:59:59.999   
8652     16502.7700   16509.4000    2466.67932 2022-11-25 18:29:59.999   
8653     16477.1300   16495.4300    2500.32611 2022-11-25 18:59:59.999   
8654     16494.0100   16515.2000    2949.47675 2022-11-25 19:29:59.999   
...             ...          ...           ...                     ...   
1222174      0.7150       0.7156  431050.00000 2022-12-13 15:29:59.999   
1222175      0.7060       0.7090  209667.00000 2022-12-13 15:59:59.999   
1222176      0.7078       0.7179  133967.70000 2022-12-13 16:29:59.999   
1222177      0.7103       0.7149  128864.20000 2022-12-13 16:59:59.999   
1222178      0.7000       0.7000  142587.90000 2022-12-13 17:29:59.999   

         Quote asset volume  Number of trades  Taker buy base asset volume  \
8650           5.557611e+07             81870                   1723.18461   
8651           6.427000e+07             91509                   2058.71422   
8652           4.073370e+07             61263                   1246.75916   
8653           4.126059e+07             59827                   1169.95110   
8654           4.872746e+07             66479                   1564.85203   
...                     ...               ...                          ...   
1222174        3.127897e+05               967                 140062.70000   
1222175        1.489892e+05               458                  33166.90000   
1222176        9.617358e+04               389                  70104.60000   
1222177        9.202996e+04               267                  59911.10000   
1222178        1.011399e+05               424                  59159.30000   

         Taker buy quote asset volume  
8650                     2.842171e+07  
8651                     3.399223e+07  
8652                     2.058911e+07  
8653                     1.930727e+07  
8654                     2.585432e+07  
...                               ...  
1222174                  1.022471e+05  
1222175                  2.363905e+04  
1222176                  5.033187e+04  
1222177                  4.276908e+04  
1222178                  4.206351e+04  

[563548 rows x 14 columns]

<p style="margin-left:2rem">➡️ As can be seen, ETH, BNB and BTC are coins traded the most on Binance</p> 

In [106]:
g_data = f_data.groupby('base').nunique()[['Symbol']].sort_values(by='Symbol', ascending=False).head(20)
fig = px.bar(g_data, x=g_data.index, y='Symbol')
fig.show()